Refactoring SCMWebhook away

hennevogel opened the pull request.

hennevogel committed:
Make SCMWebhook a concern

to be used in WorkflowRun.

hennevogel committed:
Stop tracking SCMWebhook initialization

We are going to refactor SCMWebhook away...

hennevogel committed:
Stop "validating" this concern

This will be something for TriggerWorkflowController

hennevogel committed:
Include WorkflowRunPayload concern

so we can ask WorkflowRun everything we could ask SCMWebhook before...

hennevogel committed:
Drop ScmWebhookPayloadDataExtractor

We can do this in WorkflowRun now...

hennevogel committed:
Drop unused method from ScmWebhookHeadersDataExtractor

hennevogel committed:
Introduces WorkflowRun.api_endpoint

hennevogel committed:
Split WorkflowRun payload concern SCM specific

We basically only need the SCM unspecific method public. The rest
can be private and in SCM specific concerns.

hennevogel committed:
Method to fetch the commit sha from the payload

krauselukas committed:
Set request_body variable

krauselukas committed:
Choose commit_sha based on scm event

krauselukas committed:
Add `source_repository_full_name` to workflow run payload concerns

krauselukas committed:
Add `target_repository_full_name` to workflow run payload concerns

krauselukas committed:
Add `commit_sha` to gitea workflow run payload concern

krauselukas committed:
Add `pr_number` to workflow run payload concerns

hennevogel committed:
Stop using SCMWebhook

Everything it was able to do, WorkflowRun can do now...

hennevogel committed:
Stop using TriggerControllerService::SCMExtractor

We don't need SCMWebhook anymore...

krauselukas committed:
Add `checkout_http_url` to workflow run payload concerns

krauselukas committed:
Add `tag_name` to workflow run payload concerns

krauselukas committed:
Remove callback to `extract_scm_webhook`

This method got removed.

hennevogel commented on src/api/app/models/concerns/workflow_run_payload.rb:
```diff
@@ -68,67 +60,67 @@ def ignored_push_event?
 
   def github_push_event?
-    @payload[:scm] == 'github' && @payload[:event] == 'push' && @payload.fetch(:ref, '').start_with?('refs/heads/')
+    scm_vendor == 'github' && payload[:event] == 'push' && payload.fetch(:ref, '').start_with?('refs/heads/')
```
The first part of this condition is kind of given now right? I mean the method is called `github....`

krauselukas committed:
Add `target_branch` to workflow run payload concerns

hennevogel commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Merge `SCMWebhook` into `WorkflowRun` (L)](https://trello.com/c/K9XOteaP/2468-merge-scmwebhook-into-workflowrun-l)

hennevogel commented on src/api/app/models/gitea_payload/pull_request.rb:
```diff
@@ -6,7 +6,6 @@ def payload
       commit_sha: webhook_payload.dig(:pull_request, :head, :sha),
       pr_number: webhook_payload[:number],
       source_branch: webhook_payload.dig(:pull_request, :head, :ref),
-      target_branch: webhook_payload.dig(:pull_request, :base, :ref),
```
removing only this makes little sense...

krauselukas replied:
Yes, I reverted this

krauselukas committed:
Add `object_kind` to workflow run gitlab payload concern

hennevogel commented on src/api/app/models/concerns/workflow_run_gitlab_payload.rb:
```diff
@@ -0,0 +1,53 @@
+    payload.dig('object_attributes', 'action')
+  end
+
+  def api_endpoint
```
this is duplicated from `WorkflowRunPayload` and should not be here

krauselukas replied:
Removed the duplication

hennevogel commented on src/api/app/models/concerns/workflow_run_gitea_payload.rb:
```diff
@@ -10,6 +10,12 @@ def gitea_source_repository_full_name
+  def gitea_target_repository_full_name
+    return payload.dig(:pull_request, :base, :repo, :full_name) if gitea_pull_request?
+
+    payload.dig(:repository, :full_name) # for a push event
```
We should return nil if this isn't a push event...

krauselukas replied:
Fixed

hennevogel commented on src/api/app/models/concerns/workflow_run_github_payload.rb:
```diff
@@ -25,6 +25,10 @@ def github_target_repository_full_name
   end
 
+  def github_pr_number
+    payload[:number]
```
`payload[:number]` does not exist anymore

krauselukas replied:
not? thats how we retrieved it before and it should be available through `payload[:number]`

hennevogel commented on src/api/app/models/concerns/workflow_run_gitlab_payload.rb:
```diff
@@ -43,6 +43,10 @@ def gitlab_target_branch
     payload[:after] # for a tag push event
   end
 
+  def gitlab_object_kind
```
This makes little sense as it's no universal and it's just *always* `merge_request`...

krauselukas replied:
true, its dropped now

krauselukas committed:
Use `hook_action` over `payload[:action]`

krauselukas committed:
Remove ` TriggerControllerService::SCMExtractor` class

krauselukas committed:
Split up and move scm vendor specific payload helper methods

krauselukas committed:
Make workflow run payload concerns a module instead of class

hennevogel committed:
Add more validations to WorkflowRun

hennevogel committed:
Validate WorkflowRun.request_payload is JSON

hennevogel committed:
Introduces WorkflowRun.supported_event? and .supported_action?

To be able to ask if we are going to run for this or not.

hennevogel committed:
Make use of all the new WorkflowRun validations

krauselukas committed:
Remove `scm_webhook_spec` and `scm_extractor_spec`

krauselukas committed:
Remove scm extractor related payload classes

krauselukas committed:
Remove scm webhook validator and spec

hennevogel committed:
Regernerate database_consistency todo

krauselukas committed:
Use hook_event over payload[:event]

krauselukas committed:
Adapt link_package_step_spec

krauselukas committed:
Adapt branch_package_step_spec

krauselukas committed:
Move `ALL_POSSIBLE_REQUEST_ACTIONS` constant to WorkflowRun model

Used to live in the SCMWebook model, which got removed...

eduardoj commented:
Typo in one of the commits title: "Regernerate database_consistency todo" should be "Regenerate database_consistency todo"

krauselukas committed:
Adapt configure_repositories_spec

krauselukas committed:
Adapt rebuild_package_step_spec

krauselukas committed:
Adapt set_flags_step_spec

krauselukas committed:
Adapt submit_request_step_spec

krauselukas committed:
Adapt trigger_services_spec

krauselukas committed:
Adapt workflow/step_spec.spec

krauselukas committed:
Adapt trigger_workflow_controller_spec

saraycp committed:
Adapt Workflow model spec

krauselukas committed:
Adapt order of callbacks and validations in trigger workflow controller

krauselukas committed:
Couple of fixes and additions

saraycp committed:
Adapt Workflows::ArtifactsCollector service spec

saraycp committed:
Adapt WorkflowFiltersValidator spec

hennevogel commented on src/api/app/models/concerns/workflow_run_gitea_payload.rb:
```diff
@@ -54,6 +57,22 @@ def gitea_pull_request?
+  end
+
+  def gitea_supported_event?
+    scm_vendor == 'gitea' && ALLOWED_GITHUB_EVENTS.include?(hook_event)
```
`ALLOWED_GITHUB_EVENTS`?

krauselukas replied:
Upps, yes thats the wrong one... still works because it has similar content... I will correct it

krauselukas replied:
corrected it

hennevogel commented on src/api/app/controllers/trigger_workflow_controller.rb:
```diff
@@ -70,21 +63,15 @@ def set_workflow_run
-    render_ok(data: { info: "Ignored unsupported '#{@workflow_run.hook_event}' event action: '#{@workflow_run.hook_action}'" })
+    # There are plenty of SCM events we do not handle at all (yet).
+    # Instead of throwing and errror we just ignore them.
+    if @workflow_run.errors[:hook_action].any? || @workflow_run.errors[:hook_event].any?
```
So other errors are ignored? Why so specific?

krauselukas replied:
it was already like this, we used to check if the webhook is for a supported action/event. If that was not the case we still gave back a `200` with an info that we don't support those events. If I remember correctly we did so, because people did forget to uncheck some unsupported events in the webhook configuration and the list of bad request was growing in the list of webhook deliveries on the SCM side. They thought its broken, because the webhook deliveries that actually succeeded were buried under the failing once. But since the validation for the correct event and action is now moved to the workflow_run model, I had to check for those two error cases and only then deliver a 200. In all other cases its still a 400 with the error message (two lines below)

krauselukas replied:
we did this in `skip_unsupported_actions`

hennevogel replied:
Hm okay, but you will never get here because of the return above?

hennevogel commented on src/api/spec/models/workflow_spec.rb:
```diff
@@ -321,18 +230,12 @@
     before do
       login user
+      allow(workflow_run).to receive_messages(pr_number: 1)
+      allow(workflow_run).to receive(:target_repository_full_name).and_return('iggy/test-project')
```
Why do you need to mock those?

saraycp replied:
I just need to ensure the workflow_run returns the values I need to test other stuff. I can either mock those methods or pass the request payload to the workflow_run and let it extract those values.

After implementing this I realized other specs use the second approach. So I'm going to change this for consistency and use the request payload.

saraycp replied:
Done

krauselukas committed:
Adapt `workflow_configuration_url_valid_and_accessible` validation method

We have to pass it a `WorkflowRun` object, otherwise its failing...

krauselukas committed:
Adapt token workflow spec

saraycp committed:
Adapt Workflows::YAMLToWorkflowsService spec

saraycp committed:
Remove duplication of target_package_name method

Adapt Workflow::Step#target_package_name to avoid duplication and remove
the implementation in Workflow::Step::SetFlags.

krauselukas committed:
Adapt yaml downloader and its spec

krauselukas committed:
Adapt naming of variables and add missing methods for gitea payload

krauselukas committed:
Consider gitea in workflow run model validations

krauselukas committed:
Set workflow run attributes before validation and only on create

We have to make sure the attributes are set before the validation
happens to catch eventual errors. We also have to make sure that
the attributes are only set from the request payload if they havent
been set explicitly when creating the record.

krauselukas committed:
Handle validation errors comming from workflow run

We can simply handle the active record exceptions and render the
appropriate error. Also, we can authorize first the token, there is no
need to even initialize a workflow run if a user is not permitted to run
a workflow.

Also get rid of the create_workflow_run method. The workflow creation is
moved to the 'create' method but the request_headers extraction is moved
to a specific method (to not increase complexity).

saraycp committed:
Remove no longer needed data migrations

These migrations and specs need adaptation that is not worth it.
The data migrations were applied to production long ago. It's safe to
remove it now.

- BackfillScmVendorAndHookEventInWorkflowRun. Workflow run has a new validation that doesn't allow to create it
without scm_vendor.
- RemoveUnsupportedHookEvents. The constants are moved to another model.

krauselukas committed:
Remove dropped SCMWebhook class names from zeitwerk initializer

krauselukas committed:
Drop SCMWebhook from `workflow_artifacts_per_step_component_preview`

hennevogel commented on src/api/app/models/workflow_run.rb:
```diff
@@ -26,8 +26,13 @@ class WorkflowRun < ApplicationRecord
+  validates :hook_event, inclusion: { in: ALLOWED_GITHUB_EVENTS, allow_nil: true, message: "unsupported '%{value}'" }, if: -> { scm_vendor == 'github' }
+  validates :hook_event, inclusion: { in: ALLOWED_GITLAB_EVENTS, allow_nil: true, message: "unsupported '%{value}'" }, if: -> { scm_vendor == 'gitlab' }
+  validates :hook_action, inclusion: { in: ALLOWED_PULL_REQUEST_ACTIONS, allow_nil: true, message: "unsupported '%{value}'" }, if: -> { scm_vendor == 'github' }
+  validates :hook_action, inclusion: { in: ALLOWED_MERGE_REQUEST_ACTIONS, allow_nil: true, message: "unsupported event acton '%{value}'" }, if: -> { scm_vendor == 'gitlab' }
```
What about gitea?

krauselukas replied:
Its already there, I forgot it and added it later. I guess you haven't seen it yet if you go commit by commit. To many commits...

hennevogel commented on src/api/db/data/20231023110428_remove_unsupported_hook_events.rb:
```diff
@@ -2,7 +2,8 @@
-    WorkflowRun.where.not(hook_event: [*SCMWebhookEventValidator::ALLOWED_GITHUB_AND_GITEA_EVENTS, *SCMWebhookEventValidator::ALLOWED_GITLAB_EVENTS]).destroy_all
+    allowed_events = [*WorkflowRun::ALLOWED_GITHUB_EVENTS, *WorkflowRun::ALLOWED_GITLAB_EVENTS, *WorkflowRun::ALLOWED_GITEA_EVENTS].uniq
+    WorkflowRun.where.not(hook_event: allowed_events).destroy_all
   end
```
You can also just remove this...

saraycp replied:
Removed

krauselukas committed:
Always use `merge_request` as object_kind for `_branch_request` file

Thats what the TAR SCM service expects, there are not other types
that are supported....

hennevogel commented on src/api/app/models/workflow_run.rb:
```diff
@@ -22,7 +22,7 @@ class WorkflowRun < ApplicationRecord
   ].freeze
 
-  ALL_POSSIBLE_REQUEST_ACTIONS = ['all'] + ALLOWED_PULL_REQUEST_ACTIONS + ALLOWED_MERGE_REQUEST_ACTIONS
+  ALL_POSSIBLE_REQUEST_ACTIONS = ['all'] + ALLOWED_GITHUB_PULL_REQUEST_ACTIONS + ALLOWED_MERGE_REQUEST_ACTIONS
```
What about gitea?

saraycp replied:
We can add it for clarity and in case they change in the future. Even if we know now they are duplicated from Github ones:

``` 
ALL_POSSIBLE_REQUEST_ACTIONS = (['all'] + ALLOWED_GITHUB_PULL_REQUEST_ACTIONS + ALLOWED_MERGE_REQUEST_ACTIONS + ALLOWED_GITEA_PULL_REQUEST_ACTIONS).uniq
```

hennevogel commented on src/api/app/controllers/trigger_workflow_controller.rb:
```diff
@@ -14,12 +14,19 @@ class TriggerWorkflowController < ApplicationController
     authorize @token, :trigger?
-
-    @workflow_run.save!
+    create_workflow_run
```
Those are just two lines, no need to stuff them into a method...

saraycp replied:
Moved the content of `create_workflow_run` to `create`.

saraycp committed:
Adapt SCMStatusReporter and the related spec

Now the ReportToSCMJob needs to have the workflow run set.

saraycp committed:
Introduce missing methods on payloads

saraycp committed:
Stop using subscription payload in reports

Those values are now defined in workflow_run, so we need to set the
workflow run in all the specs too.

krauselukas commented on src/api/app/services/github_status_reporter.rb:
```diff
@@ -60,9 +60,9 @@ def status_options
-      commit_sha: @event_subscription_payload[:commit_sha],
+      api_endpoint: @workflow_run.api_endpoint,
+      target_repository_full_name: @workflow_run.target_repository_full_name,
+      commit_sha: @workflow_run[:commit_sha],
```
should be `@workflow_run.commit_sha`

krauselukas replied:
adressed

hennevogel committed:
Make rubocop happy...

hellcp-work commented on src/api/app/components/workflow_run_filter_component.html.haml:
```diff
@@ -76,7 +76,6 @@
         = render partial: 'webui/shared/input', locals: { label: 'Commit', filter_item: 'commit_sha',
                                                           placeholder: 'eg. 97561db8664eaf86a1e4c7b77d5fb5d5bff6681e',
                                                           selected_input_value: @selected_filter[:commit_sha] }
-
```
Seems like this was modified in f14b9904ad045028c7b121c354f0dcda6f8f2cd7 for no reason, could it be removed from the commit?

hennevogel replied:
as you can imagine this is painful. can you just ignore it? :)

hellcp-work replied:
I can push that change myself, it's not that complex

rubhanazeem reviewed (dismissed):
LGTM! We need to monitor this closely after deployment.

hennevogel commented:
I want to deploy this to build-test.o.o before merging....

hennevogel committed:
SCMWebhook constants are now in WorkflowRun

hennevogel committed:
Clarify what the different classes abstract

hennevogel commented:
This is deployed on build-test.opensuse.org now, I've tested push/PR workflows with https://github.com/hennevogel/scmsync-package/tree/hennevogel-patch-1
